src/kudu/consensus/raft_consensus.cc

```diff
@@ -4174,13 +4174,13 @@ void RaftConsensus::HandleProxyRequest(const ConsensusRequestPB* request,
     // an error
     OpId preceding_id;
     if (request->ops_size() > 0) {
-      RET_RESPOND_ERROR_NOT_OK(queue_->log_cache()->BlockingReadOps(
-            first_op_index - 1,
-            max_batch_size,
-            read_context,
-            FLAGS_raft_log_cache_proxy_wait_time_ms,
-            &messages,
-            &preceding_id));
+      queue_->log_cache()->BlockingReadOps(
```

Comment:
Are there multiple categories of failure conditions here, and some of them
still make sense to return NOT_OK to sender?

For example if the logs have been purged by mistake on the intermediate node?

Reply:
If the logs are purged, then an error is reported and something external should clean up the instance. I think we should continue to heartbeat to destination peer.

Reply:
Discussed this some more. There are a couple of things we could try:
1. Leave it as-is in the current PR. The destination peer will eventually get removed from the config (as an error is reported by the proxy)
2. Make the proxy peer reply back with an error to the leader and not send a heartbeat. Leader will report the error and the destination will eventually be removed from the config. In the mean time, destination might start a pre-election (but will not win the election as long as the leader is stable). Since this is a rear scenario, it might be okay to let the destination start pre-election and lose it until it gets removed from the config
3. Make the proxy peer send a heartbeat to the destination, but reply back to the leader with an error. This seems little off since the destination peer might even respond with 'INVALID_TERM' and such status to the heartbeat and handling all those status correctly on the leader will be tricky.

Reply:
I agree with point 3. Giving the control to destination is the right way to go.
We need to make sure automation is not just watching the leader but also looking for purge error in peer nodes (as it happens in 1)